### PR TITLE
BUG: `read_csv` with chained fsspec TAR file and `compression="infer"`

### LINKED ISSUE #60028: BUG: read_csv with chained fsspec TAR file and compression="infer" fails with tarfile.ReadError
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import tarfile

with open("test.csv", "wt") as file:
    file.write("1;2\n3;4")
with tarfile.open("test-csv.tar", "w") as archive:
    archive.add("test.csv")

import pandas as pd

data = pd.read_csv("tar://test.csv::file://test-csv.tar", compression=None)  # works
print(data)
data = pd.read_csv("tar://test.csv::file://test-csv.tar")  # does not work
```

### Issue Description

For chained URLs, the file gets misidentified as TAR, which leads to this backtrace:

```python3
Traceback (most recent call last):
  File "/projects/ratarmount/worktrees/1/trigger-pandas-bug.py", line 12, in <module>
    data = pd.read_csv("tar://test.csv::file://test-csv.tar", compression="infer")  # does not work
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1026, in read_csv
    return _read(filepath_or_buffer, kwds)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 620, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1620, in __init__
    self._engine = self._make_engine(f, self.engine)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pandas/io/parsers/readers.py", line 1880, in _make_engine
    self.handles = get_handle(
                   ^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pandas/io/common.py", line 828, in get_handle
    handle = _BytesTarFile(
             ^^^^^^^^^^^^^^
  File "~/.local/lib/python3.12/site-packages/pandas/io/common.py", line 991, in __init__
    self.buffer: tarfile.TarFile = tarfile.TarFile.open(  # type: ignore[assignment]
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/tarfile.py", line 1842, in open
    raise ReadError(f"file could not be opened successfully:\n{error_msgs_summary}")
tarfile.ReadError: file could not be opened successfully:
- method gz: ReadError('not a gzip file')
- method bz2: ReadError('not a bzip2 file')
- method xz: ReadError('not an lzma file')
- method tar: ReadError('truncated header')
```

I have checked the source code, and the problem seems to be that the full URL is checked for ending with a TAR extension. Instead, only the last part in the chain should be checked, i.e., it should check the extension of `tar://test.csv` not `tar://test.csv::file://test-csv.tar`.

https://github.com/pandas-dev/pandas/blob/2a10e04a099d5f1633abcdfbb2dd9fdf09142f8d/pandas/io/common.py#L593-L594

### Expected Behavior

It should work without an error.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : 2a10e04a099d5f1633abcdfbb2dd9fdf09142f8d
python                : 3.12.3
python-bits           : 64
OS                    : Linux
OS-release            : 6.8.0-45-generic
Version               : #45-Ubuntu SMP PREEMPT_DYNAMIC Fri Aug 30 12:02:04 UTC 2024
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 3.0.0.dev0+1579.g2a10e04a0
numpy                 : 1.26.4
dateutil              : 2.9.0.post0
pip                   : 24.2
Cython                : 3.0.11
sphinx                : None
IPython               : 8.20.0
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
blosc                 : None
bottleneck            : None
fastparquet           : None
fsspec                : 2024.9.0
html5lib              : 1.1
hypothesis            : None
gcsfs                 : 2024.9.0post1
jinja2                : 3.1.4
lxml.etree            : 5.2.1
matplotlib            : 3.6.3
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
psycopg2              : None
pymysql               : None
pyarrow               : 15.0.2
pyreadstat            : None
pytest                : 8.3.3
python-calamine       : None
pytz                  : 2024.1
pyxlsb                : None
s3fs                  : 2024.9.0
scipy                 : 1.11.4
sqlalchemy            : None
tables                : None
tabulate              : 0.9.0
xarray                : None
xlrd                  : None
xlsxwriter            : None
zstandard             : 0.23.0
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by KevsterAmp**
- [x] closes #60028
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by KevsterAmp** [comment]
tests needs `ffspec` to be installed. What do you think @rhshadrach?

**Comment by rhshadrach** [comment]
Use the decorator `td.skip_if_no("fsspec")`.

**Comment by KevsterAmp** [comment]
CI failures seems unrelated

cc: @rhshadrach

File: pandas/tests/io/test_common.py [hunk 1]
```diff
@@ -642,6 +643,16 @@ def close(self):
...
 
 
+@td.skip_if_no("fsspec")
+def test_read_csv_chained_url_no_error():
+    tar_file = "pandas/tests/io/data/tar/test-csv.tar"
+    try:
+        pd.read_csv(f"tar://test.csv::file://{tar_file}", compression=None)
+        pd.read_csv(f"tar://test.csv::file://{tar_file}", compression="infer")
+    except Exception as e:
+        pytest.fail(e)
```
Comment (rhshadrach) [inline pandas/tests/io/test_common.py:653]: What's the benefit of using pytest.fail?

File: pandas/tests/io/test_common.py [hunk 2]
```diff
@@ -642,6 +643,16 @@ def close(self):
                 handles.created_handles.append(TestError())
 
 
+@td.skip_if_no("fsspec")
+def test_read_csv_chained_url_no_error():
+    tar_file = "pandas/tests/io/data/tar/test-csv.tar"
+    try:
+        pd.read_csv(f"tar://test.csv::file://{tar_file}", compression=None)
+        pd.read_csv(f"tar://test.csv::file://{tar_file}", compression="infer")
```
Comment (rhshadrach) [inline pandas/tests/io/test_common.py:651]: Can you check the results are as expected.

File: pandas/tests/io/test_common.py [hunk 3]
```diff
@@ -642,6 +643,16 @@ def close(self):
                 handles.created_handles.append(TestError())
 
 
+@td.skip_if_no("fsspec")
+def test_read_csv_chained_url_no_error():
+    tar_file = "pandas/tests/io/data/tar/test-csv.tar"
```
Comment (rhshadrach) [inline pandas/tests/io/test_common.py:648]: Add a comment with the issue number as the first line

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/tests/io/test_common.py (hunk 1 above)
Comment (KevsterAmp) [reply pandas/tests/io/test_common.py:653]: I was initially thinking that I need to use somekind of pytest function to validate. But it seems like it wouldn't be the case for this

**Comment by KevsterAmp** [comment]
Improved the test function to validate the dataframe inputs as well
```
(pandas-dev) kev@mac pandas % pytest pandas/tests/io/test_common.py::test_read_csv_chained_url_no_error
+ /opt/homebrew/Caskroom/miniforge/base/envs/pandas-dev/bin/ninja
[1/1] Generating write_version_file with a custom command
============================================================================================================ test session starts =============================================================================================================
platform darwin -- Python 3.10.15, pytest-8.3.3, pluggy-1.5.0
PyQt5 5.15.9 -- Qt runtime 5.15.8 -- Qt compiled 5.15.8
rootdir: /Users/kev/self/pandas
configfile: pyproject.toml
plugins: localserver-0.0.0, qt-4.4.0, cov-5.0.0, hypothesis-6.115.5, cython-0.3.1, anyio-4.6.2.post1, xdist-3.6.1
collected 1 item

pandas/tests/io/test_common.py .

------------------------------------------------------------------------------------------ generated xml file: /Users/kev/self/pandas/test-data.xml ------------------------------------------------------------------------------------------
============================================================================================================ slowest 30 durations ============================================================================================================
0.01s call     pandas/tests/io/test_common.py::test_read_csv_chained_url_no_error

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
```

File: pandas/tests/io/test_common.py [hunk 4]
```diff
@@ -642,6 +643,21 @@ def close(self):
...
+@td.skip_if_no("fsspec")
+def test_read_csv_chained_url_no_error():
+    # GH 60100
+    tar_file = "pandas/tests/io/data/tar/test-csv.tar"
+
+    x = pd.read_csv(f"tar://test.csv::file://{tar_file}", compression=None)
+    y = pd.read_csv(f"tar://test.csv::file://{tar_file}", compression="infer")
+
+    x_to_json_expected_output = '{"1;2":{"0":"3;4"}}'
+    y_to_json_expected_output = '{"1;2":{"0":"3;4"}}'
```
Comment (rhshadrach) [inline pandas/tests/io/test_common.py:655]: Can you just create the expected DataFrame and use `tm.assert_frame_equal`. Conversion JSON is lossy.

**Comment by KevsterAmp** [comment]
CI looks unrelated

**Comment by rhshadrach** [comment]
> CI looks unrelated

The test that's failing is the one being added here.

https://github.com/pandas-dev/pandas/actions/runs/11700807499/job/32585578297?pr=60100#step:8:83

**Comment by KevsterAmp** [comment]
> The test that's failing is the one being added here.

Any ideas to fix it? Since the CI was `minimum versions` I assumed that it was just outdated version of Tarfile or pandas

**Comment by mroeschke** [comment]
Looks like a bug with the version of fsspec being tested (`TarContainedFile` in fsspec didn't define flush). You'll want to specify a minimum version in `skip_if_no` (2023.1.0)

File: pandas/tests/io/test_common.py [hunk 5]
```diff
@@ -642,6 +643,21 @@ def close(self):
...
 
 
+@td.skip_if_no("fsspec")
+def test_read_csv_chained_url_no_error():
+    # GH 60100
+    tar_file_path = "pandas/tests/io/data/tar/test-csv.tar"
+    chained_file_url = f"tar://test.csv::file://{tar_file_path}"
+
+    result_a = pd.read_csv(chained_file_url, compression=None, sep=";")
+    result_b = pd.read_csv(chained_file_url, compression="infer", sep=";")
```
Comment (mroeschke) [inline pandas/tests/io/test_common.py:653]: Could you use `pytest.mark.parametrize` over the `compression` parameter

**Review by mroeschke — APPROVED** [review]

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @KevsterAmp

**Pull request merged**